pycroft/model/user.py

```diff
@@ -341,6 +344,46 @@ def permission_level(self):
                    default=0)
 
 
+manager.add_function(
+    User.__table__,
+    ddl.Function(
+        'user_room_change_update_history', [],
+        'trigger',
+        """
+        BEGIN
+            IF old.room_id IS DISTINCT FROM new.room_id THEN
+                IF old.room_id IS NOT NULL THEN
+                    /* User was living in a room before, history entry must be ended */
+                    UPDATE "room_history_entry" SET ends_at = CURRENT_TIMESTAMP
+                        WHERE room_id = old.room_id AND user_id = new.id AND ends_at IS NULL;
+                END IF;
+
+                IF new.room_id IS NOT NULL THEN
+                    /* User moved to a new room. history entry must be created */
+                    INSERT INTO "room_history_entry" (user_id, room_id, begins_at)
+                        /* We must add one second so that the user doesn't have two entries
+                           for the same timestamp */
+                        VALUES(new.id, new.room_id, CURRENT_TIMESTAMP);
+                END IF;
+            END IF;
+            RETURN NULL;
+        END;
+        """,
+        volatility='volatile', strict=True, language='plpgsql'
+    )
+)
+
+manager.add_trigger(
+    User.__table__,
+    ddl.Trigger(
+        'user_room_change_update_history_trigger',
+        User.__table__,
+        ('UPDATE', 'INSERT'),
```

Comment:
What happens on room deletion (the only thing taking action should be the cascades)?
I do realize we're unlikely to ever do that.